sql/core/src/main/java/org/apache/spark/sql/sources/v2/writer/WriteTask.java

```diff
@@ -0,0 +1,66 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.sources.v2.writer;
+
+import java.io.Serializable;
+
+import org.apache.spark.annotation.Experimental;
+import org.apache.spark.annotation.InterfaceStability;
+import org.apache.spark.sql.Row;
+import org.apache.spark.sql.catalyst.encoders.ExpressionEncoder;
+import org.apache.spark.sql.catalyst.encoders.RowEncoder;
+import org.apache.spark.sql.catalyst.expressions.UnsafeRow;
+import org.apache.spark.sql.types.StructType;
+
+/**
+ * A write task that is responsible for writing the data for each input partition.
+ */
+public abstract class WriteTask implements Serializable {
+  private ExpressionEncoder<Row> encoder;
+
+  /**
+   * This method will be called before writing data, users can overwrite this method and put
+   * initialization logic here. Note that, when overwriting this method, you have to call
+   * `super.initialize` to correctly do some internal setup for this write task.
+   */
+  public void initialize(StructType schema) {
+    encoder = RowEncoder.apply(schema);
+  }
+
+  /**
+   * The actual write logic should be implemented here. To correctly implement transaction,
+   * implementations should stage the writing or have a way to rollback.
+   */
+  public abstract void write(Row row);
```

Comment:
`InternalRow` is not a public interface, `Row` is. So the main write interface should take `Row`, and leave the `UnsafeRow` write interface for advanced users.

Reply:
BTW if you overwrite the `UnsafeRow` write interface, you can just throw exception in this `Row` write interface.

Reply:
Yeah, I think I would push my previous point then. Really the write(Row) method should be a protected method that the user is expected to override (if spark is never going to call it). However, the fact that everyone who wants to implement an efficient datasource is forced to make this code throw or something (this is not javadoc'd!) makes me think that this is dodgy code.

I would prefer a

```
interface WriteTask implements Serializable {
    void initialize(StructType schema);
    void write(InternalRow row);
    void WriterCommitMessage commit();
}
```

```
abstract class RowWriteTask implements WriteTask {
    protected abstract void write(Row row);

    // impls of initialize and write(InternalRow).
}
```

which to me is a much more clear separation of concerns - the person who can accept InternalRows does not need to care that UnsafeRows exist

Reply:
One other thing - I feel like the 'initialize' method is literally present to construct the converter.

In practice, StructType is serializable and we know it when producing the tasks. I think that we should not have an initialize method and should instead have it be a constructor arg of RowWriteTask.

The other thing is related to a concern I had wrt ReadTask - where it's typically a lot easier to ship the config required to construct something than it is to ship the actual writer objects (and I don't really want an initialize method - it makes it harder to guarantee that my objects are in a safe state).

What would be awesome is if WriteTask were more of a factory class, and if it could be made to return something that contains the 'write, commit, abort' methods (maybe have WriteTask have a method which returns a Writer?).

Reply:
Explicit init is extremely important in life cycle management.